feat: update search index when course content is updated [FC-0040]

rpenido opened the pull request:
## Description
This PR implements updating the search index with content metadata whenever they change

## More Infomation
- Closes https://github.com/openedx/modular-learning/issues/196
- Depends on:
  - https://github.com/openedx/edx-platform/pull/34310

## Testing Instructions
- Run your local stack on this branch
- Make sure you have `meilisearch` setup locally, follow the setup instructions here https://github.com/open-craft/tutor-contrib-meilisearch
- Make sure you also have some sample taxonomies/tags, you can add some from here: https://github.com/open-craft/taxonomy-sample-data
- Run `tutor dev run cms bash` and `./manage.py cms reindex_studio`
- View the resulting index at http://meilisearch.local.edly.io:7700/ (see [tutor-contrib-meilisearch README](https://github.com/open-craft/tutor-contrib-meilisearch) for how to get the API key to log in)
- Go back to studio, and add/update XBlocks from a course
- Go back to meilisearch dashboard, and confirm that the XBlocks docs have been updated to reflect the changes you made
- Repeat and confirm the above for both units and blocks
- Also, try to create a Content Library with some components and check if the index is updated accordingly.

**NOTE:** Meilisearch seems to cache queries along with their results in the frontend, so if you simply search it might 
show you stale data (network tab shows no requests), especially if you're searching for the same query. Make sure you refresh the Meilisearch dashboard (http://meilisearch.local.edly.io:7700/) and then perform the search.

___
Private ref: [FAL-3690](https://tasks.opencraft.com/browse/FAL-3690)

openedx-webhooks commented:
Thanks for the pull request, @rpenido! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

yusuf-musleh commented on openedx/core/djangoapps/content/search/api.py:
```diff
@@ -0,0 +1,370 @@
+
+    # Connect to Meilisearch
+    if not settings.MEILISEARCH_ENABLED:
+        raise RuntimeError("MEILISEARCH_ENABLED is not set - search functionality disabled.")
```
Should this check be at the top so it raises the error instead of returning the client when its not enabled:

```suggestion
    # Connect to Meilisearch
    if not settings.MEILISEARCH_ENABLED:
        raise RuntimeError("MEILISEARCH_ENABLED is not set - search functionality disabled.")
    
    if _MEILI_CLIENT is not None:
        return _MEILI_CLIENT
```

yusuf-musleh commented on openedx/core/djangoapps/content/search/handlers.py:
```diff
@@ -0,0 +1,76 @@
+        recursive=True,  # Update all children because the breadcrumb may have changed
+        update_metadata=True,
+        update_tags=False,
+    )
```
Why is `update_tags` set to `False` here? Is it because we are going to implement updating tags in the index separately in https://github.com/openedx/modular-learning/issues/197 ?

rpenido replied:
Yes! This handler is called when we receive an `XBLOCK_UPDATED` event. If this block already has tags, I think the index will still be up to date with its information.

The `upsert_xblock_index_doc` did not replace the document in the index; it only updates it with the new fields (https://www.meilisearch.com/docs/reference/api/documents#add-or-update-documents).

For the tagging data, I was planning to mirror the implementation we did here:
- Create an `OBJECT_TAGGED` event that is dispatched when we call the `tag_object` api.
- Create a new handler here that only updates the tags for the object and leaves the metadata as it is (`{recursive=False, update_metadata: False, updat_tags=True}`);

But I'm open to new ideas if you think otherwise! :smile:

bradenmacdonald replied:
Yes, as long as we are careful not to overwrite the `tags` field with an empty list, this is a good approach. Generally the tags haven't changed when this event is called, so there's no reason to update them.

yusuf-musleh replied:
Got it! Thanks for the clarification, and I agree I think it's a good approach. I'll follow it when working on https://github.com/openedx/modular-learning/issues/197

yusuf-musleh reviewed (commented):
Great work on this @rpenido ! I realized after reviewing that a lot of the changes on this PR diff are from @bradenmacdonald 's PR since this PR wasn't updated with upstream yet. None the less I think the code overall is really well structured and easy to follow!

Just had a small question/nit.

However in terms of testing, I'm not sure if its ready or not, but after briefly tried I was faced with some errors (maybe because the PR needs to be updated with upstream), so I just reviewed the code for now. @rpenido Let me know when its ready for another review/testing it out.

rpenido commented:
> I realized after reviewing that a lot of the changes on this PR diff are from @bradenmacdonald 's PR since this PR wasn't updated with upstream yet.

We have this temp PR (https://github.com/open-craft/edx-platform/pull/645). It could be easier to review before we get #34310 merged.

> However in terms of testing, I'm not sure if its ready or not, but after briefly tried I was faced with some errors

Sorry for that @yusuf-musleh! I probably messed up something when I merged the changes from @bradenmacdonald.

I will let you know when this is ready for testing again (probably for a final review).

Also, I started writing some tests for the new API, but I'm not happy with them. As meilisearch is an external service, I need to mock many functions and only really test if the external API is called with the expected parameters. Maybe we should fire up a Meileisearch instance during the tests to actually test the feature (as we do with external services, like the database), but I'm not sure if this is the right time to do this (and I don't know how to do this in our CI suit either).

I plan to continue with the current approach (mocking and checking calls) for the API and the handlers to get some coverage, but I think we should revisit this if the prototype becomes the default search engine. Let me know if you have a better idea for this!

yusuf-musleh commented:
> As meilisearch is an external service, I need to mock many functions and only really test if the external API is called with the expected parameters.

@rpenido I think this approach is fine, since we can make the assumption that Meilisearch will behave as expected after we manually tested it. However I can see the value of having integration tests to actually test that it behaves as expected in case there are breaking/unexpected changes on the Meilisearch side. Although that could be avoided by using the same working version vs updating to the latest version every time.

> I think we should revisit this if the prototype becomes the default search engine

I agree, I think this can come at a later stage. As you suggested, we can potentially follow the footsteps of the [mysqldb](https://github.com/openedx/edx-platform/blob/master/.github/workflows/docker-compose.yml.mysqldbdump) when the time comes.

bradenmacdonald commented:
@rpenido I agree that we should eventually run Meilisearch during tests. However, it's probably not necessary to implement that for this first studio search project which is considered experimental. Maybe you could add that to the ADR that I wrote? That for the experiment, we won't use Meilisearch during tests, but we expect to add that in the future if we move forward with replacing Elasticsearch completely.

rpenido commented on openedx/core/djangoapps/content/search/tests/test_handlers.py:
```diff
@@ -0,0 +1,174 @@
+                "breadcrumbs": [{"display_name": "Test Course"}, {"display_name": "Updated Sequential"}],
+                "content": {}
+            }
+        ])
```
@yusuf-musleh, I liked how you used `copy` in your PR. I will update this code with it after the review.

rpenido commented:
Hi @yusuf-musleh ! There is a test failing here, but I can't find the logs. I will look into it further tomorrow.

Edit: fixed 5a6beafc86276aafa9a02ca895f49b9d691a2905

yusuf-musleh reviewed (commented):
@rpenido Great work, the code is coming along quite well! The code looks good to me, however while testing the library aspects, I noticed a few issues:

- When I create a new empty library, nothing gets added to the index, is that expected?
- When I add a component/block to the library that gets added to the index, as expected, when I edit the blocks, eg: the name or the problem type (for problem blocks) the index does not get updated. However If I change the name of the library, thats when the blocks under it get updated in the index with the latest information. It's probably just a missing condition on how/when the docs should be updated in the index.

rpenido commented:
Thank you for the review @yusuf-musleh!

>     * When I create a new empty library, nothing gets added to the index, is that expected?

I think that it is expected. I commented on the [first PR](https://github.com/openedx/edx-platform/pull/34310#issuecomment-2010729981) but I think it got lost in the review comments.
@bradenmacdonald, could you confirm that we are not indexing the Library and the Course entities?

>     * When I add a component/block to the library that gets added to the index, as expected, when I edit the blocks, eg: the name or the problem type (for problem blocks) the index does not get updated. However If I change the name of the library, thats when the blocks under it get updated in the index with the latest information. It's probably just a missing condition on how/when the docs should be updated in the index.

This is because the `LIBRARY_BLOCK_UPDATED` event is not dispatched when we edit an library XBlock. Not sure if we will fix it now or later.

rpenido commented:
@yusuf-musleh I want to rebase and squash this PR, but you will need to rebase it and merge conflicts again after that too (it will happen anyway when this goes upstream). Let me know when you think it is a good time to do it!

bradenmacdonald commented:
@rpenido  @yusuf-musleh  We aren't indexing the courses/libraries themselves. So it's expected that a newly created library won't appear in the results.

yusuf-musleh commented:
> I want to rebase and squash this PR, but you will need to rebase it and merge conflicts again after that too (it will happen anyway when this goes upstream). Let me know when you think it is a good time to do it!

@rpenido  Go for it! I'll handle the conflicts on my end, no worries.

rpenido committed:
feat: update search index when course content is updated

rpenido commented:
> > I want to rebase and squash this PR, but you will need to rebase it and merge conflicts again after that too (it will happen anyway when this goes upstream). Let me know when you think it is a good time to do it!
> 
> @rpenido Go for it! I'll handle the conflicts on my end, no worries.

Thank you @yusuf-musleh! Rebase done!

rpenido committed:
feat: add celery retry to search tasks

rpenido commented:
Hi @bradenmacdonald ! This is ready for CC review

bradenmacdonald commented on openedx/core/djangoapps/content/search/documents.py:
```diff
@@ -196,33 +196,58 @@ def _tags_for_content_object(object_id: UsageKey | LearningContextKey) -> dict:
 
-def searchable_doc_for_library_block(metadata: lib_api.LibraryXBlockMetadata) -> dict:
+def searchable_doc_for_library_block(
+    xblock_metadata: lib_api.LibraryXBlockMetadata, include_metadata: bool = True, include_tags: bool = True
```
Nit: it's kinda hard to read this line - would be easier with one arg per line.

But actually, when do we _not_ want to include "metadata"? Shouldn't that just always be True? I feel like this is adding some complexity that we don't need, but maybe I'm missing something.

In fact, thinking about it more, it's probably fine to just always update tags too, but I guess that's fine either way.

rpenido replied:
> But actually, when do we _not_ want to include "metadata"? 

When we tag the object, we are not updating "metadata" (i.e. breadcrumbs, name, block type, etc..). We could simplify and always update everything if you think this suits best.

bradenmacdonald replied:
Yes, but that can be a separate function.

```python
def update_tags(usage_key):
    doc = { Fields.id: meili_id_from_opaque_key(usage_key) }
    doc.update(_tags_for_content_object(usage_key))
    _wait_for_meili_task(client.index(studio_index). update_documents([doc]))
```

I think since it's so simple and it works for both course/library blocks, it's better to keep the "update tags" as a separate function, rather than making it a parameter to both `searchable_doc_for_library_block` and `searchable_doc_for_course_block`.

CC @yusuf-musleh

rpenido replied:
Done: d2c3c539b6c615c1b18120b0cf55a703dbd84d83

bradenmacdonald commented on openedx/core/djangoapps/content/search/documents.py:
```diff
@@ -196,33 +196,58 @@ def _tags_for_content_object(object_id: UsageKey | LearningContextKey) -> dict:
+    }
+
+    if include_metadata:
+        doc.update(_fields_from_block(block))
```
I wouldn't really call this "metadata" - this includes the text of the actual content of the block. Maybe "main_fields" or something. But again I think it's simpler to just always include this, without making it optional.

bradenmacdonald commented on openedx/core/djangoapps/content/search/api.py:
```diff
@@ -0,0 +1,486 @@
+    _wait_for_meili_tasks(tasks)
+
+
+def generate_user_token(user):
```
We might have other indexes in the future, so this should be called `generate_user_token_for_studio_search` or something.

rpenido replied:
Done: 1c498b8b14cc4f27e43d7d23a83cd075c29656d

bradenmacdonald commented on openedx/core/djangoapps/content/search/api.py:
```diff
@@ -0,0 +1,486 @@
+
+    tasks = []
+    if current_rebuild_index_name:
+        # If there is a rebuild in progress, the document will also be added to the new index.
```
```suggestion
        # If there is a rebuild in progress, the document will also be deleted from the new index.
```

bradenmacdonald commented on openedx/core/djangoapps/content/search/api.py:
```diff
@@ -0,0 +1,486 @@
+    tasks = []
+    if current_rebuild_index_name:
+        # If there is a rebuild in progress, the document will also be added to the new index.
+        tasks.append(client.index(current_rebuild_index_name).delete_document(meili_id_from_opaque_key(usage_key)))
```
Will this raise an error if the document doesn't exist in the new index? (We don't want it to - it should just silently ignore if the document doesn't exist.)

rpenido replied:
If the document doesn't exist, it doesn't throw an error. Calling `delete_document` only puts the task in the queue (returning 202).

Getting the task status, we got the following (also, no error):
```
uid=247
index_uid='tutor_studio_content'
status='succeeded'
type='documentDeletion'
details={'providedIds': 1, 'deletedDocuments': 0, 'originalFilter': None}
error=None
canceled_by=None
duration='PT0.002798645S'
enqueued_at=datetime.datetime(2024, 4, 5, 22, 15, 1, 921564)
started_at=datetime.datetime(2024, 4, 5, 22, 15, 1, 923094)
finished_at=datetime.datetime(2024, 4, 5, 22, 15, 1, 925892)
```

bradenmacdonald commented on openedx/core/djangoapps/content/search/api.py:
```diff
@@ -0,0 +1,486 @@
+        pass
+
+    if status_cb is None:
+        status_cb = nop
```
If no `status_cb` is passed, we should log using the `log` function from line 35.

Also please document the `status_cb` parameter in the docstring; it's not super clear what it's for.

rpenido replied:
Done: 8e64b137df3c03cd092b7b39ad93197f2123ad31

bradenmacdonald commented on openedx/core/djangoapps/content/search/api.py:
```diff
@@ -0,0 +1,486 @@
+        pass
+
+    if status_cb is None:
+        status_cb = nop
```
If no `status_cb` is passed, we should log using the `log` function from line 35.

rpenido replied:
Done: 814d0a6680c02789ddf2d8d0b8d42bc5f693ea3f

bradenmacdonald commented on openedx/core/djangoapps/content/search/api.py:
```diff
@@ -0,0 +1,486 @@
+    client = _get_meilisearch_client()
+    current_status = client.get_task(info.task_uid)
+    while current_status.status in ("enqueued", "processing"):
+        time.sleep(1)
```
Let's reduce this to `0.5` as I suspect it will often need to wait less than a second. Also, the docstring should mention that this should only ever be used in management commands or celery tasks. We don't want to block a web worker thread using `sleep`. Annoying limitation of python.

rpenido replied:
Done: bb7ac372660fd3a22bde3e592d426a092a332584

bradenmacdonald commented on openedx/core/djangoapps/content/search/api.py:
```diff
@@ -0,0 +1,486 @@
+_MEILI_CLIENT = None
+_MEILI_API_KEY_UID = None
+
+LOCK_EXPIRE = 5 * 60  # Lock expires in 5 minutes
```
I don't think 5 minutes is enough... An index rebuild could take hours on a large instance.

When we call `cache.add()` we should have a very long timeout on the lock - probably 24 hours. And instead of blocking and waiting for the lock for LOCK_EXPIRE seconds here, we should just immediately throw an error.

For context, rebuilding the index is a rare thing done only by a sysadmin after a new installation or major upgrade or something, and not expected to happen most of the time.

rpenido replied:
Thank you for the context!

Updated here: c8d265165772bf474d4d8f3feac405f8e5fb4580

bradenmacdonald commented on openedx/core/djangoapps/content/search/api.py:
```diff
@@ -0,0 +1,486 @@
+if hasattr(settings, "MEILISEARCH_INDEX_PREFIX"):
+    INDEX_NAME = settings.MEILISEARCH_INDEX_PREFIX + STUDIO_INDEX_NAME
+else:
+    INDEX_NAME = STUDIO_INDEX_NAME
```
I find `STUDIO_INDEX_NAME` and `INDEX_NAME` a bit confusing here.

Perhaps rename `STUDIO_INDEX_NAME` to `STUDIO_INDEX_SUFFIX` and rename `INDEX_NAME` to `STUDIO_INDEX_NAME` because what you've called `INDEX_NAME` _is_ the final name of the studio index.

rpenido replied:
Done: e3bc29d31f59e86f4f4b632babb05ea2a42ac24f

For now, I'm using the constants inside the functions, but we can easily use the index name as a parameter in the future.

bradenmacdonald commented on openedx/core/djangoapps/content/search/api.py:
```diff
@@ -0,0 +1,486 @@
+        tasks.append(client.index(current_rebuild_index_name).update_documents(docs))
+    tasks.append(client.index(INDEX_NAME).update_documents(docs))
+
+    _wait_for_meili_tasks(tasks)
```
These lines (from `if not docs` to `_wait_for_meili_tasks`) seem to be repeated in most of these methods. Maybe factor them out to a helper method?

rpenido replied:
Done: 0c20792af0f4afe7c916675c00f01d39672fd242

Way leaner now! Thank you!

bradenmacdonald reviewed (commented):
@rpenido Sorry for the delay on my review.

I think @ormsbee will also want to take a look at this, maybe on Monday once you've been able to address my comments.

rpenido committed:
fix: use log.info if status callback not defined

rpenido committed:
fix: update docstring and reduce sleep time

rpenido committed:
refactor: remove update_metadata and update_tags parameters

rpenido committed:
refactor: rename generate_user_token to generate_user_token_for_studio_search

rpenido committed:
docs: fix docstring

Co-authored-by: Braden MacDonald <mail@bradenm.com>

rpenido committed:
refactor: rename index name constants

rpenido committed:
refactor: create _update_index_docs helper

rpenido committed:
fix: increase rebuild lock time and throws if rebuild already in progress

rpenido committed:
docs: add docstring for status_cb

rpenido committed:
style: fix pylint

bradenmacdonald reviewed (approved):
Great, thanks for those changes @rpenido !

@yusuf-musleh  Would you mind just quickly testing this again to make sure it's working?

@ormsbee Would you like to review this? (Up to you)

yusuf-musleh commented:
> Would you mind just quickly testing this again to make sure it's working?

@rpenido @bradenmacdonald I've tested it again, it's working well as expected. There's just the library content issue from before, where the `LIBRARY_BLOCK_UPDATED` is not firing when library blocks are updated (hence the index is not updated). However it seems like its fine for now and it'll be addressed in a separate ticket/issue?

bradenmacdonald commented:
Thanks @yusuf-musleh. Yeah we don't need to worry too much about library blocks for now.

rpenido committed:
Merge branch 'master' into rpenido/fal-3690-update-search-index-when-course-content-is-changed

rpenido committed:
Merge branch 'master' into rpenido/fal-3690-update-search-index-when-course-content-is-changed

bradenmacdonald committed:
Merge branch 'master' into rpenido/fal-3690-update-search-index-when-course-content-is-changed

bradenmacdonald commented:
This is ready to merge but we're waiting on some unrelated issues with the NodeJS 18 upgrade to be solved first.

rpenido commented:
Hi @bradenmacdonald ! I added this commit (612f32ff6355fa48f5af2c8fae50503706b70dc1) here.
There are some old libraries (before the recent module store refactor) that crash our reindex.

I deleted these libraries locally, and the code was working. In the sandbox, the error appeared again, so I think it would be better to skip the library if we found an error instead of crashing the reindex.

rpenido committed:
fix: error in library crash reindex

```File"/openedx/edx-platform/openedx/core/djangoapps/content_libraries/api.py",line
617, in get_library_componentslearning_package.id,AttributeError:
'NoneType' object has no attribute'id'```

bradenmacdonald commented:
Thanks @rpenido - that's a great fix.

bradenmacdonald commented:
@rpenido I was just about to merge this, but there are some conflicts with the "permissions" PR that I just merged. Could you please fix them?

rpenido committed:
Merge branch 'master' into rpenido/fal-3690-update-search-index-when-course-content-is-changed

rpenido committed:
fix: fix code and tests after merge

rpenido commented:
> @rpenido I was just about to merge this, but there are some conflicts with the "permissions" PR that I just merged. Could you please fix them?

Working on it. The code changed a lot, so I will need to manually test it to make sure I did the merge right.

bradenmacdonald commented:
@rpenido Take your time - it's important to get it right :)
Thanks.

rpenido committed:
style: fix pylint

rpenido committed:
feat: update meilisearch configuration needed for tags filter

Co-authored-by: Braden MacDonald <braden@opencraft.com>

rpenido committed:
fix: conditional import

rpenido committed:
style: fix pylint

rpenido commented:
This is ready for review and merge @bradenmacdonald. It also includes the changes from your PR:
- https://github.com/openedx/edx-platform/pull/34534#pullrequestreview-2007408802

It may have impacted your PR @yusuf-musleh.

openedx-webhooks commented:
@rpenido 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/34391&entry.1484655318=https://github.com/rpenido&entry.752974735=2024-03-19+18:38&entry.1917517419=2024-04-18+16:53&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
